Google.Api.Gax.Grpc/ClientBuilderBase.cs

```diff
@@ -422,15 +438,22 @@ protected async virtual Task<ChannelCredentials> GetChannelCredentialsAsync(Canc
         private protected virtual ChannelBase CreateChannel(string endpoint, ChannelCredentials credentials) =>
             EffectiveGrpcAdapter.CreateChannel(endpoint, credentials, GetChannelOptions());
 
-        private class DelegatedTokenAccess : ITokenAccess
+        private class DelegatedTokenAccess : ITokenAccessWithHeaders
         {
             private readonly Func<string, CancellationToken, Task<string>> _tokenAccessMethod;
+            private readonly string _quotaProject;
 
-            internal DelegatedTokenAccess(Func<string, CancellationToken, Task<string>> tokenAccessMethod) =>
-                _tokenAccessMethod = tokenAccessMethod;
+            internal DelegatedTokenAccess(Func<string, CancellationToken, Task<string>> tokenAccessMethod, string quotaProject) =>
+                (_tokenAccessMethod, _quotaProject) = (tokenAccessMethod, quotaProject);
 
             public Task<string> GetAccessTokenForRequestAsync(string authUri, CancellationToken cancellationToken) =>
                 _tokenAccessMethod(authUri, cancellationToken);
+
+            public async Task<AccessTokenWithHeaders> GetAccessTokenWithHeadersForRequestAsync(string authUri = null, CancellationToken cancellationToken = default)
+            {
+                string token = await _tokenAccessMethod(authUri, cancellationToken).ConfigureAwait(false);
+                return new AccessTokenWithHeaders.Builder { QuotaProject = _quotaProject }.Build(token);
```

Comment:
If this is going to be called on every request, and *normally* QuotaProject isn't set, might it make sense to optimize for that case. So something like:

```csharp
return QuotaProject is null
    ? token
    : new AccessTokenWithHeaders.Builder { QuotaProject = _quotaProject }.Build(token);
```

Alternatively, could we keep a single `AccessTokenWithHeader.Builder` for all requests? I can't remember whether it's safe to reuse that.

Reply:
The `AccessTokenWithHeader.Builder.Build` method is not thread safe right now, so reusing the builder is not an option. I can make it thread safe, but I'd rather do that later, rerelease Google.Apis, and the come back here if that's OK.

I will optimize for `QuotaProject` not being set though by only using the builder if it is. Otherwise I can use the `AccessTokenWithHeader` constructor.

Let me know if this is fine.

Reply:
Yup, that sounds reasonable.